fix: show error message depend on the language preference

StevenKKC opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ [#18203](https://github.com/Expensify/App/issues/18203)
PROPOSAL: [#18203(COMMENT)](https://github.com/Expensify/App/issues/18203#issuecomment-1530100343)


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Go Preference
2. Set offline
3. Change the language to Spanish
4. Click on new chat with invalid number, example +15005550111
5. Send messages
6. Go back to Preference, change the language to English
7. Go to the same chat again and set online
8. Verify that the error messages is shown in English when the language was already changed to English.
- [X] The error message should be shown in English

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
Same as Tests step.

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
In web and desktop version, there's a console error that is not related to this PR. A console error is generated in src/styles/getTooltipStyles.js and src/libs/BootSplash/index.js, and display error screen.
I think should open an issue for it.

- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/131836113/09036250-33b1-47aa-b77c-77b6bddf3444

</details>

<details>
<summary>Mobile Web - Chrome</summary>


https://github.com/Expensify/App/assets/131836113/24a35d27-4c66-4704-99b8-bb58c73fe87e



</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/131836113/e2da1f28-3890-4f8f-96f3-73ff76031d04

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/131836113/750f66a7-5c76-44da-9946-1706c87e7265

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/131836113/7e012156-ba23-49e3-b0c9-422dd7b14b55

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/131836113/7303f33f-cc40-4653-8849-beb99f783fa8

</details>

StevenKKC committed:
fix: show error message depend on the language preference

fedirjh reviewed (changes requested):
@StevenKKC As this is your first time contribution , you need to sign the CLA , check this https://github.com/Expensify/App/pull/19324#issuecomment-1555329877 . Also please update the checklist , the template should’ve been maintained . There are some lint error that you should fix as well.

Please include all screenshots for all platforms.

StevenKKC commented:
I have read the CLA Document and I hereby sign the CLA

StevenKKC committed:
fix lint error & add getOptimisticErrors utility function

StevenKKC commented:
@fedirjh Updated. 
Sorry, I'm not pay attention to many things.
I have fixed lint errors, and according to your suggestion, added a utility function. Please review again.
Android screenshot and mobile chrome screenshot, I'll upload a little later.

StevenKKC commented:
@fedirjh Uploaded the missing screenshots and tests step.

fedirjh reviewed (dismissed):
Looks good to me.

cc @chiragsalian

🎀👀🎀 C+ reviewed

StevenKKC commented:
@fedirjh Thanks for your review.

@chiragsalian Could you please review my PR?

StevenKKC committed:
Merge branch 'main' into stevenkkc-fixErrorMessage

StevenKKC commented:
@fedirjh @chiragsalian Because of conflict, I have merged main branch into my branch.

StevenKKC committed:
remove Localize import in Policy.js

chiragsalian commented on src/libs/ErrorUtils.js:
```diff
@@ -37,6 +38,17 @@ function getAuthenticateErrorMessage(response) {
+ */
+function getOptimisticErrors(error) {
+    // eslint-disable-next-line rulesdir/prefer-localization
+    return ({[DateUtils.getMicroseconds()]: error});
```
unnecessary outer brackets `()`.

StevenKKC committed:
fix lint error in ErrorUtils.js

StevenKKC committed:
fix lint error in Session/index.js

StevenKKC commented:
@fedirjh @chiragsalian I have fixed lint errors.

chiragsalian reviewed (changes requested):
Code mostly LGTM, left some minor suggestions

StevenKKC committed:
rewrite utility function and remove unnecessary lint comment

StevenKKC committed:
change function comment.

StevenKKC commented:
@chiragsalian Updated files according to your suggestion.

StevenKKC commented:
@chiragsalian @fedirjh All checks have passed. What can I do for this now?

StevenKKC commented:
@chiragsalian Could you please review the changes that you requested? I hope to merge this PR as soon as possible.

chiragsalian commented on src/libs/ErrorUtils.js:
```diff
@@ -37,6 +38,16 @@ function getAuthenticateErrorMessage(response) {
+ * @return {Object} - Object to be optimistically saved to Onyx
+ *
+ */
+function getOptimisticErrors(error) {
```
Discussed this a bit with my team. We just found the method name to be a bit confusing. The method doesn't really give us optimisticData. It actually gives us failureData.

Can you rename this to,
```
getMicroSecondOnyxError
```

chiragsalian commented on src/libs/ErrorUtils.js:
```diff
@@ -37,6 +38,16 @@ function getAuthenticateErrorMessage(response) {
 }
 
+/**
+ * Method used to generate errors object optimistically.
```
Once we rename the method i think this description could be updated.
```suggestion
 * Method used to get an error object with microsecond as the key.
```

chiragsalian commented on src/libs/ErrorUtils.js:
```diff
@@ -37,6 +38,16 @@ function getAuthenticateErrorMessage(response) {
+/**
+ * Method used to generate errors object optimistically.
+ * @param {String} error - error key or message to be saved
+ * @return {Object} - Object to be optimistically saved to Onyx
```
```suggestion
 * @return {Object}
```

chiragsalian reviewed (changes requested):
Yup i wanted to discuss this solution with my team so it took me a bit.

Looks like you missed a spot [here](https://github.com/Expensify/App/blob/2a786a71bc37004824a1012c1fd2fe4225ebf68b/src/libs/actions/ReimbursementAccount/errors.js#L45) and [here](https://github.com/Expensify/App/blob/2a786a71bc37004824a1012c1fd2fe4225ebf68b/src/libs/actions/SignInRedirect.js#L52).

Other than that code LGTM. i left some minor suggestion just to improve readability.

chiragsalian reviewed (changes requested):
Yup i wanted to discuss this solution with my team so it took me a bit.

Looks like you missed a spot [here](https://github.com/Expensify/App/blob/2a786a71bc37004824a1012c1fd2fe4225ebf68b/src/libs/actions/ReimbursementAccount/errors.js#L45) and [here](https://github.com/Expensify/App/blob/2a786a71bc37004824a1012c1fd2fe4225ebf68b/src/libs/actions/SignInRedirect.js#L52).

Other than that code LGTM. i left some minor suggestion just to improve readability.

StevenKKC committed:
change ErrorUtils function name and comment

StevenKKC commented:
@chiragsalian I have updated files according to your request. Could you please review again?

chiragsalian reviewed (changes requested):
I think you missed addressing [this comment](https://github.com/Expensify/App/pull/19324#pullrequestreview-1440322120) of mine.
Specifically this one,
> Looks like you missed a spot [here](https://github.com/Expensify/App/blob/2a786a71bc37004824a1012c1fd2fe4225ebf68b/src/libs/actions/ReimbursementAccount/errors.js#L45) and [here](https://github.com/Expensify/App/blob/2a786a71bc37004824a1012c1fd2fe4225ebf68b/src/libs/actions/SignInRedirect.js#L52).

StevenKKC committed:
fix missed error message

StevenKKC commented:
@chiragsalian @fedirjh Updated.

StevenKKC commented:
@chiragsalian All checks have passed. Please let me know what I should do.

chiragsalian reviewed (approved):
LGTM

chiragsalian commented:
I'm going ahead and merging since fedi already approved it earlier and his most recent comment was minor.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5074571096) to staging by https://github.com/chiragsalian in version: 1.3.18-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5074571096) to staging by https://github.com/chiragsalian in version: 1.3.18-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|cancelled 🔪 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5074571096) to staging by https://github.com/chiragsalian in version: 1.3.18-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5087951817) to production by https://github.com/yuwenmemon in version: 1.3.18-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

StevenKKC commented:
I have checked the Deploy checklist (https://github.com/Expensify/App/issues/19565) and the issue #19651 (https://github.com/Expensify/App/issues/19651).

But I think it's not related to this PR at all.

When user change the language preference, the following code will be executed.

https://github.com/Expensify/App/blob/0f1188de4d7b69ecb2b16485e9f03d92f54afe84/src/libs/actions/App.js#L85-L112

In the offline mode (or the user have already changed the force offline mode to online, but the app is still in the offline status, as you can see in your video), API.write only change the local Onyx data, and the request API that update the language preference does not send to the server.

![screenshot](https://github.com/Expensify/App/assets/131836113/96eaadd7-daa8-47de-be6b-3d7f6778189a)

After send the comment, the server will send the response error message and the Onyx data that is stored in the server to the app, and these data will contain the original language preference, as you can see in the below screenshot.

![screenshot1](https://github.com/Expensify/App/assets/131836113/55fa3fc6-8a69-4014-8a68-da7b1b01455d)

So, after app receive the server response and merge the server's Onyx data, the language preference will be changed to the previous state, that is Spanish, and the error message is shown in Spanish.

That is, this PR is working as expected.

luacmartins commented on src/libs/actions/BankAccounts.js:
```diff
@@ -81,7 +81,7 @@ function getVBBADataForOnyx() {
                 value: {
                     isLoading: false,
                     errors: {
-                        [DateUtils.getMicroseconds()]: Localize.translateLocal('paymentsPage.addBankAccountFailure'),
```
We're passing `'paymentsPage.addBankAccountFailure'` to the utility method, but all it's doing is storing that key as the error instead of translating it. So this is what the error looks like:

![Screenshot 2023-06-02 at 10 04 35 AM](https://github.com/Expensify/App/assets/22219519/6fc15a92-46ee-4316-969d-4c1be3e7b872)

but it should be:

`1685721717986000: "An unexpected error occurred while trying to add your bank account. Please try again."`

luacmartins replied:
We're only calling `translateIfPhraseKey` in DotIndicator, but there are other places displaying the error text, e.g. Forms, that don't use DotIndicator.
